internal/northbound/upload.go

```diff
@@ -167,28 +186,38 @@ func (g *Server) UploadCatalogEntities(ctx context.Context, req *catalogv3.Uploa
 }
 
 func (u *uploadSession) processUploadSession(ctx context.Context, tx *generated.Tx) error {
-	orderedSpecs, err := u.loadYamlSpecs()
+	// Turn the uploads into independent filesets. Each tarball will be a fileset, and
+	// any raw files will be collected into a fileset.
+	fileSets, err := u.loadFileSets()
 	if err != nil {
 		return err
 	}
 
-	for _, d := range orderedSpecs {
-		switch d.SpecSchema {
-		case upload.DeploymentPackageType:
-			err = u.loadDeploymentPackage(ctx, tx, d)
-		case upload.DeploymentPackageLegacyType:
-			err = u.loadDeploymentPackage(ctx, tx, d)
-		case upload.ApplicationType:
-			err = u.loadApplication(ctx, tx, d)
-		case upload.RegistryType:
-			err = u.loadRegistry(ctx, tx, d)
-		case upload.ArtifactType:
-			err = u.loadArtifact(ctx, tx, d)
-		default:
-			return nberrors.NewInvalidArgument(nberrors.WithMessage("uploaded file %s: unhandled type %s", d.FileName, d.SpecSchema))
-		}
+	// Process each fileset, load its yaml, sort, and add to the catalog.
+	for _, fileSet := range fileSets {
+		orderedSpecs, err := u.loadYamlSpecs(fileSet)
 		if err != nil {
-			return nberrors.NewInvalidArgument(nberrors.WithError(err), nberrors.WithMessage("uploaded file %s: %v", d.FileName, err))
+			return err
+		}
+
+		for _, d := range orderedSpecs {
+			switch d.SpecSchema {
+			case upload.DeploymentPackageType:
+				err = u.loadDeploymentPackage(ctx, tx, d)
+			case upload.DeploymentPackageLegacyType:
+				err = u.loadDeploymentPackage(ctx, tx, d)
+			case upload.ApplicationType:
+				err = u.loadApplication(ctx, tx, d, fileSet) // application uses the fileSet to lookup profiles
+			case upload.RegistryType:
+				err = u.loadRegistry(ctx, tx, d)
+			case upload.ArtifactType:
+				err = u.loadArtifact(ctx, tx, d)
+			default:
+				return nberrors.NewInvalidArgument(nberrors.WithMessage("uploaded file %s: unhandled type %s", d.FileName, d.SpecSchema))
```

Comment:
Would it be good idea to ignore the file and be permissive?

Reply:
done